presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java

```diff
@@ -66,33 +64,27 @@
     private int rowPosition = -1;
     private int inPageIndex = -1;
 
-    private QueryResultRows(Session session, List<Type> types, List<Column> columns, List<Page> pages)
-    {
-        this(session, types, columns, pages, null);
-    }
-
-    private QueryResultRows(Session session, List<Type> types, List<Column> columns, List<Page> pages, Consumer<Throwable> exceptionConsumer)
+    private QueryResultRows(Session session, Optional<List<ColumnAndType>> columns, List<Page> pages, Consumer<Throwable> exceptionConsumer)
     {
         this.session = requireNonNull(session, "session is null").toConnectorSession();
-        this.types = requireNonNull(types, "types is null");
         this.columns = requireNonNull(columns, "columns is null");
         this.pages = new ArrayDeque<>(requireNonNull(pages, "pages is null"));
         this.exceptionConsumer = Optional.ofNullable(exceptionConsumer);
         this.totalRows = countRows(pages);
         this.currentPage = this.pages.pollFirst();
         this.supportsParametricDateTime = session.getClientCapabilities().contains(ClientCapabilities.PARAMETRIC_DATETIME.toString());
 
-        verify(this.types.size() == this.columns.size(), "columns and types sizes mismatch");
+        verify(totalRows == 0 || (totalRows > 0 && columns.isPresent()), "data present without columns and types");
```

Comment:
This is a new verification, it should be added as a separate commit.
(when splitting, make sure to split the tests accordingly too)